riscv/src/asm_translate/mod.rs

```diff
@@ -0,0 +1,575 @@
+use std::{
+    cell::Cell,
+    collections::{BTreeMap, BTreeSet, HashSet},
+};
+
+use itertools::Itertools;
+use parser::RiscParser;
+use powdr_asm_utils::{
+    ast::{BinaryOpKind, UnaryOpKind},
+    data_parser,
+    data_storage::store_data_objects,
+    parser::parse_asm,
+    reachability::{self, symbols_in_args},
+    utils::{argument_to_escaped_symbol, argument_to_number, expression_to_number},
+    Architecture,
+};
+use powdr_number::FieldElement;
+
+use crate::{
+    code_gen::{self, Args, FunctionKind, MemEntry, Register, RiscVProgram, SourceFileInfo},
+    Runtime,
+};
+
+mod disambiguator;
+mod parser;
+
+type Statement = powdr_asm_utils::ast::Statement<Register, FunctionKind>;
+type Argument = powdr_asm_utils::ast::Argument<Register, FunctionKind>;
+type Expression = powdr_asm_utils::ast::Expression<FunctionKind>;
+
+struct AsmProgram {
+    file_ids: Vec<(i64, String, String)>,
+    mem_entries: Cell<Option<Vec<MemEntry>>>,
+    statements: Vec<Statement>,
+}
+
+const START_FUNCTION: &str = "__stack_setup";
+
+impl<'a> RiscVProgram<'a> for AsmProgram {
+    type InstructionArgs = [Argument];
+    type Label = &'a str;
+
+    fn source_files_info(&self) -> impl Iterator<Item = SourceFileInfo> {
+        self.file_ids.iter().map(|(id, dir, file)| SourceFileInfo {
+            id: *id as u32,
+            dir,
+            file,
+        })
+    }
+
+    fn initial_mem(&self) -> impl Iterator<Item = MemEntry> {
```

Comment:
i think the usage of this method will be more clear as `take_initial_mem(&mut)` and dropping the `Cell` usage

Reply:
Done.